@xen-orchestra/audit-core/src/index.js

```diff
@@ -0,0 +1,176 @@
+      algorithm,
+    })}`
+  }
+
+  @synchronizedFns
+  async add(subject, event, data) {
+    const record = {
+      data,
+      event,
+      previousHash: (await this.getData()).hashes.pop(),
+      subject,
+      time: Date.now(),
+    }
+    record.id = this._createHash(record)
+    await this._storage.add(record.id, record)
+    return record
+  }
+
+  async checkIntegrity(startHash, endHash) {
+    const record = await this._storage.get(endHash)
+    if (record === undefined) {
+      throw new Error(
+        `The record associated to the hash ${hash} doesn't exists`
+      )
+    }
+
+    if (
+      endHash !==
+      this._createHash(
+        omit(record, 'id'),
+        ID_TO_ALGORITHM[endHash.slice(1, endHash.indexOf('$', 1))]
+      )
+    ) {
+      throw new Error(`The hash ${hash} not correspond to the stored record`)
+    }
+
+    if (endHash === startHash) {
+      return
+    }
+
+    if ((endHash = record.previousHash) === undefined) {
+      throw new Error(
+        `The records between ${startHash} and ${hash} are missing`
+      )
+    }
+
+    return this.checkIntegrity(startHash, endHash)
+  }
+
+  async generateSecuredInterval() {
+    const { hashes } = await this.getData()
+    if (hashes.length === 0) {
+      throw new Error('Empty storage')
+    }
+
+    const retention = this._retention
+    if (retention !== undefined) {
+      const diff = hashes.length - retention
+      if (diff > 0) {
+        hashes.splice(0, diff)
+      }
+    }
+
+    const startHash = hashes[0]
+    const endHash = hashes.pop()
+    await this.checkIntegrity(startHash, endHash)
+
+    const storage = this._storage
+    const result = {
+      startHash: {
+        hash: startHash,
+        time: (await storage.get(startHash)).time,
+      },
+      endHash: {
+        hash: endHash,
+        time: (await storage.get(endHash)).time,
+      },
+    }
+    await this._storage.update('', globalData => ({
+      ...globalData,
+      ...result,
+    }))
+
+    return result
+  }
+
+  async getData() {
+    const { '': globalData = {}, ...records } =
+      (await this._storage.getAll()) ?? {}
+    const sortedRecords = sortBy(records, 'time')
+    return {
+      '': globalData,
+      hashes: sortedRecords.map(({ id }) => id),
+      records: sortedRecords,
+    }
+  }
+
+  async gc() {
+    const { startHash } = (await this._storage.get('')) ?? {}
+    if (startHash === undefined) {
+      return
+    }
+
+    const storage = this._storage
+    let record = await storage.get(startHash.hash)
+    if (record === undefined) {
+      throw new Error(
+        `the records corresponding to the start hash (${startHash.hash}) doesn't exists`
+      )
+    }
+
+    const promises = []
+    while ((record = await storage.get(record.previousHash)) !== undefined) {
+      promises.push(storage.del(record.id))
+    }
+
+    return Promise.all(promises)
+  }
+
+  @synchronizedFns
+  async reWriteHashes() {
```

Comment:
Why do we need this?

Reply:
This method will be useful when a user will want to repair the hash chain after a deletion of a sensitive data or a modification of the records.

Reply:
Don't expose this method, if we want to propose a way to delete or replace a record, we should create explicit methods for it, not a a method which will silently rewrite the whole history.

Reply:
What if a chain is broken and the user identify the issue and want to repair it?

Reply:
As optimization, this method can takes optionally a start hash to not re-write all the hashes.

Reply:
> What if a chain is broken and the user identify the issue and want to repair it?

Does this need exist today?

> As optimization, this method can takes optionally a start hash to not re-write all the hashes.

It's not a question of optimization, it's a question of control and trustability in the audit.